specs/agents/tracing-instrumentation-messaging.md

```diff
@@ -1,26 +1,87 @@
+occurring within a traced transaction.
+
+![publish](uml/publish.svg)
+
+### Trace Context
+
+if the messaging system exposes an API for sending additional message properties/metadata, it
+SHOULD be used to propagate the [Trace Context](https://www.w3.org/TR/trace-context/) of the
+`messaging` span, to continue the [distributed trace](tracing-distributed-tracing.md).
+
+## Message reception/consumption
+
+Message reception can be divided into two types, **passive** and **active**. 
+
+### Passive message reception
+
+Passive message reception typically involves a listener or callback subscribed
+to a queue/topic/subscription that is called once a message or batch of messages
+is available, and passed the message or batch of messages for processing.
+
+Passive reception SHOULD be captured as a `messaging` transaction that starts when
+a message or batch of messages is received for processing, and ends when the
+processing of the message or batch of messages finishes.
+
+### Active message reception
+
+Active message reception typically involves initiating the reception of a
+message or batch of messages from a queue/topic, on demand. This might be a
+blocking or non-blocking operation, and for the purposes of this spec, is
+collectively referred to as polling.
+
+When message polling is performed within a traced transaction, it SHOULD be
+captured in a `messaging` span with a name indicating a poll action. A
+span SHOULD be created irrespective of whether the operation returns any messages.
+
+If message polling is performed when there is no active transaction, it
+can be the initiating event for a message processing flow. Our goal in this
+scenario is to capture the message processing flow as a `messaging` transaction
+that SHOULD start after the polling operation exits **and** MUST only be started
+if the polling operation returned a message or batch of messages. Capturing
+the message processing flow in a transaction in this manner may be unfeasible if
+the processing flow is not implemented within a well defined API.
+
+### Trace Context
+
+When message reception is captured as a `messaging` transaction, if the messaging
+system exposes an API for sending additional message properties/metadata, it
+SHOULD be checked for the presence of [Trace Context](https://www.w3.org/TR/trace-context/).
+If Trace Context is present, it SHOULD be propagated to the `messaging` transaction
+to continue the [distributed trace](tracing-distributed-tracing.md).
+
+If message reception results in a batch of messages that are processed in a
+message processing flow and hence captured in a `messaging` transaction, it may be
+possible that each message in the batch has its own Trace Context. In this
+scenario, it is not currently possible to propagate a Trace Context to the `messaging`
+transaction, since there a multiple contexts present. It may be possible to capture
+these in future [(issue: #122)](https://github.com/elastic/apm/issues/122).
+
```

Comment:
```suggestion
possible that each message in the batch has its own Trace Context. In this
scenario, a transaction SHOULD be started and ended for each message 
handling wherever possible. For example, the Java agent instruments the 
Kafka consumer batch iterator so that a transaction is started whenever 
a message is retrieved from the batch and ended either when the next message 
is retrieved, of when the iterator is depleted (i.e. `iterator.hasNext()` returns 
`false`).
If this is not possible, the agent SHOULD create a single transaction for the 
batch. It may be possible to capture multiple parent contexts for such 
batch-transactions in the future through [span links](https://github.com/elastic/apm/issues/122).

```

Reply:
Yes, it can go into a separate section, but currently in this section there is a sentence saying: 
```
In this scenario, it is not currently possible to propagate a Trace Context to 
the `messaging` transaction, since there are multiple contexts present.
```
This contradicts my addition. So the intention was, in addition to introducing the alternative way of dealing with batches, to say that the preferable way is to not create a single transaction for the batch, e.g. to propagate the trace context to separate transactions, and only whenever this is not possible, we will create a single transaction, ignoring the multiple trace contexts until we have span links in place.

Reply:
I commented here that it should still change, but just reread in the entire context and all makes sense, thanks!